botocore/credentials.py

```diff
@@ -791,6 +821,157 @@ def _assume_role_base_kwargs(self, config):
+
+    def _retrieve_temp_credentials(self):
+        logger.debug("Retrieving credentials via AssumeRoleWithSaml.")
+        response = self._create_client().assume_role_with_saml(
+            **self._assume_role_base_kwargs(self._get_role_config_values()))
+        creds = self._create_creds_from_response(response)
+        return creds, response
+
+    def _assume_role_base_kwargs(self, config):
+        if config.get('saml_authentication_type')=='form':
+            if not config.get('saml_username'):
+                config['saml_username'] = self.username_prompter("Username: ")
+            mapping = {
+                "adfs": {
+                    "username": "ctl00$ContentPlaceHolder1$UsernameTextBox",
+                    "password": "ctl00$ContentPlaceHolder1$PasswordTextBox"},
+                }.get(config.get('saml_provider'),
+                      {"username": "username", "password": "password"})
+            assertion = SamlFormsBasedAuthenticator().authenticate(
+                config['saml_endpoint'],
+                {mapping["username"]: config['saml_username'],
+                 mapping["password"]: self.password_prompter("Password: ")},
+                verify=config.get('saml_verify_ssl') != 'false')
+        else:
+            raise ValueError("Unsupported saml_authentication_type: %s"
+                             % config.get('saml_authentication_type'))
+        if not assertion:
+            raise ValueError('Login failed: SAML assertion not found')
+        idp_roles = self._parse_roles(assertion)
+        if not idp_roles:
+            raise ValueError('Identity provider provides no role.')
+        role = self.role_selector(config.get('role_arn'), idp_roles)
+        if not role:
+            raise ValueError('Unable to choose role "%s" from %s' % (
+                config.get('role_arn'), [r['RoleArn'] for r in idp_roles]))
+        role['SAMLAssertion'] = assertion
+        return role
+
+    def _parse_roles(self, assertion):
+        attribute = '{urn:oasis:names:tc:SAML:2.0:assertion}Attribute'
+        attr_value = '{urn:oasis:names:tc:SAML:2.0:assertion}AttributeValue'
+        awsroles = []
+        root = ET.fromstring(base64.b64decode(assertion))
+        for attr in root.getiterator(attribute):
+            if attr.get('Name')=='https://aws.amazon.com/SAML/Attributes/Role':
+                for value in attr.getiterator(attr_value):
+                    parts = value.text.split(',')
+                    # Deals with "role_arn,pricipal_arn" or its reversed order
+                    if 'saml-provider' in parts[0]:
+                        role = {'PrincipalArn': parts[0], 'RoleArn': parts[1]}
+                    else:
+                        role = {'PrincipalArn': parts[1], 'RoleArn': parts[0]}
+                    awsroles.append(role)
+        return awsroles
+
+
+class SamlFormsBasedAuthenticator(object):
+    def authenticate(self, endpoint, params, verify=True):
```

Comment:
Docstring.  I'm following this call from the assume role provider and don't quite understand what `params` is suppose to be.